jtag serial drive write function loses characters when rx ring buffer… (IDFGH-10925)

jjsch-dev opened the pull request:
upon examining the code, I noticed that the [usbjtag_tx_char_via_driver](https://github.com/espressif/esp-idf/blob/32472536715d674c160a2565795895e0273f8cde/components/vfs/vfs_usb_serial_jtag.c#L406) function contains a comparison that checks whether an immediate send operation succeeds. However, the return value from [usb_serial_jtag_write_bytes](https://github.com/espressif/esp-idf/blob/32472536715d674c160a2565795895e0273f8cde/components/driver/usb_serial_jtag/usb_serial_jtag.c#L158) is always 1, which prevents the intended logic to try blocking for 50 mS, but also a return that always skyp the bloking try.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-idf?pullRequest=12119) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-idf?pullRequest=12119) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-idf?pullRequest=12119) it.</sub>

ginkgm commented:
usb_serial_jtag_write_bytes never returning 0 is a real issue. We will fix this.

The API will return 0 when write fails. Change this to -1 is a breaking change that we can't do.

From the VFS side, losing data when ringbuffer full is a expected behavior. We can't detect the connection to host, so we use the fifo overflow as indication.

jjsch-dev commented:
Hi @ginkgm, thanks for the explanation, if you want I'll close the PR.

jjsch-dev commented:
Hi @ginkgm, I modified the return of usb_serial_jtag_write_bytes to return 0 on failure and the size on success, and reverted the change to usbjtag_tx_char_via_driver.

chipweinberger commented:
@jjsch-dev thanks for creating the PR!

jjsch-dev commented:
@chipweinberger, thanks for the comment.

mythbuster5 commented:
@jjsch-dev  Yes, this time changes is absolutely correct, I already have an internal PR to fix this, and that will be merged soon. Thanks for your contribution.

jjsch-dev commented:
@mythbuster5, thanks for the confirmation, I will wait the merge.
